# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

- **Consider a significance level of 5% for all tests.**

## Libraries

In [38]:
import numpy as np
import pandas as pd
from scipy import stats
from scipy.stats import chi2_contingency

# Challenge 1 - Independent Sample T-tests

In this challenge, we will be using the Pokemon dataset. Before applying statistical methods to this data, let's first examine the data.

- To load the data, run the code below.

In [2]:
# Run this code:
pokemon = pd.read_csv('pokemon.csv')

In [3]:
pokemon = pokemon.set_index('#')

- Let's start off by looking at the `head` function in the cell below.

In [4]:
# Your code here:
pokemon.head()

Name Type 1  Type 2  Total  HP  Attack  Defense  Sp. Atk  \
#                                                                              
1              Bulbasaur  Grass  Poison    318  45      49       49       65   
2                Ivysaur  Grass  Poison    405  60      62       63       80   
3               Venusaur  Grass  Poison    525  80      82       83      100   
3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123      122   
4             Charmander   Fire     NaN    309  39      52       43       60   

   Sp. Def  Speed  Generation  Legendary  
#                                         
1       65     45           1      False  
2       80     60           1      False  
3      100     80           1      False  
3      120     80           1      False  
4       50     65           1      False

In [5]:
pokemon.info()

<class 'pandas.core.frame.DataFrame'>
Index: 800 entries, 1 to 721
Data columns (total 12 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Name        800 non-null    object
 1   Type 1      800 non-null    object
 2   Type 2      414 non-null    object
 3   Total       800 non-null    int64 
 4   HP          800 non-null    int64 
 5   Attack      800 non-null    int64 
 6   Defense     800 non-null    int64 
 7   Sp. Atk     800 non-null    int64 
 8   Sp. Def     800 non-null    int64 
 9   Speed       800 non-null    int64 
 10  Generation  800 non-null    int64 
 11  Legendary   800 non-null    bool  
dtypes: bool(1), int64(8), object(3)
memory usage: 75.8+ KB


The dataset consists of information on 800 Pokémon, each represented by 13 attributes. These attributes include basic details like Name, Type 1, and Type 2, as well as various stats such as Total, HP, Attack, Defense, Sp. Atk, Sp. Def, and Speed. Additionally, it provides information on each Pokémon's Generation and whether it is a Legendary Pokémon, indicated by a boolean value. Notably, the Type 2 column contains some missing values (only 414 entries are non-null), which is expected, as not all Pokémon have a secondary type.

- The first thing we would like to do is compare the legendary Pokemon to the regular Pokemon. To do this, we should examine the data further. What is the count of legendary vs. non legendary Pokemons?

In [6]:
# Your code here:
print(f'There are {len(pokemon[pokemon['Legendary'] == True])} Legendary Pokémon in this dataset')
print(f'There are {len(pokemon[pokemon['Legendary'] != True])} Non-Legendary Pokémon in this dataset')

There are 65 Legendary Pokémon in this dataset
There are 735 Non-Legendary Pokémon in this dataset


- Compute the mean and standard deviation of the total points for both legendary and non-legendary Pokemon.

In [8]:
pokemon.groupby('Legendary')['Total'].agg(['mean', 'std']).reset_index().round(2)

Legendary    mean     std
0      False  417.21  106.76
1       True  637.38   60.94

The computation of the mean might give us a clue regarding how the statistical test may turn out; However, it certainly does not prove whether there is a significant difference between the two groups.

In the cell below, use the `ttest_ind` function in `scipy.stats` to compare the the total points for legendary and non-legendary Pokemon. Since we do not have any information about the population, assume the variances are not equal.

**Null Hypothesis (H₀):**  
- $\bar{x}_{\text{Legendary}} = \bar{x}_{\text{non-Legendary}}$  
(means from Legendary and non-Legendary stats are equal)

**Alternative Hypothesis (H₁):**  
- $\bar{x}_{\text{Legendary}} \neq \bar{x}_{\text{non-Legendary}}$  
(means from Legendary and non-Legendary stats are not equal)

In [9]:
# Your code here:
t_stat, p_value = stats.ttest_ind(pokemon[pokemon['Legendary'] == True]['Total'],
                                 pokemon[pokemon['Legendary'] == False]['Total'])

print(f'T-test result: {round(t_stat, 2)}')
print(f'p-value: {p_value} (basically {round(p_value, 2)})')

T-test result: 16.39
p-value: 3.0952457469652825e-52 (basically 0.0)


- What do you conclude from this test? Write your conclusions below.

The t-test result of 16.39 and the p-value of basically 0.0 show a strong difference between the two groups. We absolutely reject (H₀). It’s clear that Legendary Pokémon have significantly higher base stats compared to Non-Legendary ones. Who would guess that, right?

How about we try to compare the different types of pokemon? In the cell below, list the types of Pokemon from column `Type 1` and the count of each type.

In [10]:
# Your code here:
pokemon['Type 1'].value_counts().reset_index().sort_values(by='count', ascending=False)

Type 1  count
0      Water    112
1     Normal     98
2      Grass     70
3        Bug     69
4    Psychic     57
5       Fire     52
6   Electric     44
7       Rock     44
10     Ghost     32
9     Ground     32
8     Dragon     32
11      Dark     31
12    Poison     28
13     Steel     27
14  Fighting     27
15       Ice     24
16     Fairy     17
17    Flying      4

Since water is the largest group of Pokemon, compare the mean and standard deviation of water Pokemon to all other Pokemon.

In [11]:
# Your code here:
print(f'Water Pokémon Stats:\n{pokemon[pokemon['Type 1'] == 'Water']['Total'].agg(['mean', 'std']).round(2)}')
print(f'Remaining Pokémon Stats:\n{pokemon[pokemon['Type 1'] != 'Water']['Total'].agg(['mean', 'std']).round(2)}')

Water Pokémon Stats:
mean    430.46
std     113.19
Name: Total, dtype: float64
Remaining Pokémon Stats:
mean    435.86
std     121.09
Name: Total, dtype: float64


- Perform a hypothesis test comparing the mean of total points for water Pokemon to all non-water Pokemon. Assume the variances are equal. 

**Null Hypothesis (H₀):**
- $ \bar{x}_{\text{water}} = \bar{x}_{\text{non-water}} $  
(The sample means are equal)

**Alternative Hypothesis (H₁):**  
- $ \bar{x}_{\text{water}} \neq \bar{x}_{\text{non-water}} $  
(The sample means are different)

In [12]:
# Your code here:
t_stat, p_value = stats.ttest_ind(pokemon[pokemon['Type 1'] == 'Water']['Total'],
                                 pokemon[pokemon['Type 1'] != 'Water']['Total'])

print(f'T-test result: {round(t_stat, 2)}')
print(f'p-value: {p_value} (basically {round(p_value, 2)})')

T-test result: -0.44
p-value: 0.6587140317488793 (basically 0.66)


- Write your conclusion below:

The t-test result of -0.44 and the p-value of 0.66 show that there’s no significant difference between Water Pokémon and the rest of the Pokémon when it comes to total stats. With a p-value this high, we can’t reject the null hypothesis, so there’s no clear evidence to suggest that Water Pokémon have different average stats compared to the others.

# Challenge 2 - Matched Pairs Test

In this challenge we will compare dependent samples of data describing our Pokemon. Our goal is to see whether there is a significant difference between each Pokemon's defense and attack scores. **Our hypothesis is that the defense and attack scores are equal**. In the cell below, import the `ttest_rel` function from `scipy.stats` and compare the two columns to see if there is a statistically significant difference between them.

---

When it says "our hypothesis is that the defense and attack scores are equal," it might sound like it’s referring to the alternative hypothesis. However, it’s actually referring to the **null hypothesis**.

In hypothesis testing, the **null hypothesis** is the starting assumption. We assume that there’s **no difference** between the groups, which in this case means the defense and attack scores are equal. The **alternative hypothesis** would then say that there’s a **difference** between the two.

So, in this case, "our hypothesis" is actually setting up the **null hypothesis**, even though the wording seems to point to the alternative. The null hypothesis would be:

 $$ H_0: \bar{x}_{\text{Atk}} = \bar{x}_{\text{Def}} $$

And the alternative hypothesis would be:

 $$ H_1: \bar{x}_{\text{Atk}} \neq \bar{x}_{\text{Def}} $$

The confusion comes from the phrasing. When referring to "our hypothesis," it typically means the alternative hypothesis, but in this case, it’s the null hypothesis being assumed.

---

In [13]:
from scipy.stats import ttest_rel

In [14]:
t_stat, p_value = ttest_rel(pokemon['Attack'], pokemon['Defense'])

print(f'T-test: {round(t_stat, 2)}')
print(f'p-value: {p_value} (basically {round(p_value, 2)})')

T-test: 4.33
p-value: 1.7140303479358558e-05 (basically 0.0)


- Describe the results of the test in the cell below.

The t-statistic of 4.33 means there's a noticeable difference between the attack and defense scores. With a p-value close to 0.0, we reject the null hypothesis.

We are also curious about whether therer is a significant difference between the mean of special defense and the mean of special attack. Perform the hypothesis test in the cell below. 

- $H_0: \bar{x}_{\text{(Sp. Atk)}} = \bar{x}_{\text{(Sp. Def)}}$
- $H_1: \bar{x}_{\text{(Sp. Atk)}} \neq \bar{x}_{\text{(Sp. Def)}}$

In [15]:
t_stat, p_value = ttest_rel(pokemon['Sp. Atk'], pokemon['Sp. Def'])

print(f'T-test: {round(t_stat, 2)}')
print(f'p-value: {round(p_value, 2)}')

T-test: 0.85
p-value: 0.39


In [16]:
pokemon[['Attack', 'Defense']].agg(['mean', 'std']).reset_index().round(2)

index  Attack  Defense
0  mean   79.00    73.84
1   std   32.46    31.18

In [17]:
pokemon[['Sp. Atk', 'Sp. Def']].agg(['mean', 'std']).reset_index().round(2)

index  Sp. Atk  Sp. Def
0  mean    72.82    71.90
1   std    32.72    27.83

- Describe the results of the test in the cell below.

The t-statistic of 0.85 means the difference between the Sp. Atk and Sp. Def scores is pretty small. With a p-value of 0.39, we don't have enough evidence to reject the null hypothesis, so there's no significant difference between the two.

$$ I - HONESTLY - DON'T - GET - HOW $$

As you may recall, a two sample matched pairs test can also be expressed as a one sample test of the difference between the two dependent columns.

- Import the `ttest_1samp` function and perform a one sample t-test of the difference between defense and attack. Test the hypothesis that the difference between the means is zero. Confirm that the results of the test are the same.

In [18]:
from scipy.stats import ttest_1samp

In [19]:
t_test, p_value = stats.ttest_1samp((pokemon['Defense'] - pokemon['Attack']), 0)

print(f'T-test: {round(t_stat, 2)}')
print(f'p-value: {round(p_value, 2)}')

T-test: 0.85
p-value: 0.0


In [20]:
t_test, p_value = stats.ttest_1samp((pokemon['Sp. Def'] - pokemon['Sp. Atk']), 0)

print(f'T-test: {round(t_stat, 2)}')
print(f'p-value: {round(p_value, 2)}')

T-test: 0.85
p-value: 0.39


# Bonus Challenge - The Chi-Square Test

The Chi-Square test is used to determine whether there is a statistically significant difference in frequencies. In other words, we are testing whether there is a relationship between categorical variables or rather when the variables are independent. This test is an alternative to Fisher's exact test and is used in scenarios where sample sizes are larger. However, with a large enough sample size, both tests produce similar results. Read more about the Chi Squared test [here](https://en.wikipedia.org/wiki/Chi-squared_test).

- In the cell below, create a contingency table using `pd.crosstab` comparing whether a Pokemon is legenadary or not and whether the Type 1 of a Pokemon is water or not.

$$
H_0: \text{No association between Legendary status and Type 1 being Water}
$$

$$
H_1: \text{There is an association between Legendary status and Type 1 being Water}
$$

In [35]:
# Your code here:
legxwat = pd.crosstab(pokemon['Legendary'], pokemon['Type 1'] == 'Water')
legxwat

Type 1     False  True 
Legendary              
False        627    108
True          61      4

- Perform a chi-squared test using the `chi2_contingency` function in `scipy.stats`. You can read the documentation of the function [here](https://docs.scipy.org/doc/scipy-0.15.1/reference/generated/scipy.stats.chi2_contingency.html).

In [42]:
# Your code here:
chi2_stat, p_value, dof, expected = chi2_contingency(legxwat)

print(f'chi2: {round(chi2_stat, 2)}')
print(f'p-value: {round(p_value, 2)}')

chi2: 2.94
p-value: 0.09


- Based on a 95% confidence, should we reject the null hypothesis?

Shockingly, there is no statistical evidence supporting that any relationship between a Pokémon being Legendary and having it's primary Type as Water.

---

The overall conclusion of the analysis is that Pokémon is an incredible game that combines strategy, adventure, and nostalgia in a way few others can. Whether you’re battling gym leaders, exploring vast worlds, or simply collecting creatures, the game offers something for everyone. It’s a perfect mix of fun and challenge, where players can test their skills and build unique teams. Pokémon isn’t just a game; it’s a journey that brings people together and encourages curiosity, making it a must-play for anyone who loves to dive into new worlds.

---

**Disclosure**: any compliments are directed solely to the core games (this means Pokémon Anime and Pokémon GO are OUT! They're not cool!)